js/noms/src/spec.js

```diff
@@ -0,0 +1,315 @@
+ * accessors have the same mutability semantics as the `Database` interface,
+ * which is mutable. For example, if the underlying `Database` commits a
+ * dataset, then calls to `Spec.dataset()` will return the new one.
+ *
+ * For example:
+ *  - a database spec might be `mem` or `https://demo.noms.io/cli-tour`.
+ *  - a dataset spec might be `mem::photos` or `https://demo.noms.io/cli-tour::sf-crime`.
+ *  - a path spec might be `http://demo.noms.io::foo.bar` or `http://demo.noms.io::#abcdef.bar`
+ *
+ * See https://github.com/attic-labs/noms/blob/master/doc/spelling.md.
+ */
+export default class Spec {
+  _spec: string;
+  _opts: ?SpecOptions;
+  _protocol: string;
+  _databaseName: string;
+
+  // Lazily created.
+  _database: ?Database;
+
+  // Exists if the spec was constructed through `forDataset()`.
+  _datasetName: ?string;
+
+  // Exists if the spec was constructed through `fromValue()` or `pin()`.
+  _path: ?AbsolutePath;
+
+  constructor(spec: string, databaseSpec: string, opts: ?SpecOptions) {
+    this._spec = spec;
+    this._opts = opts;
+    const [protocol, databaseName] = parseDatabaseSpec(databaseSpec);
+    this._protocol = protocol;
+    this._databaseName = databaseName;
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a database.
+   */
+  static forDatabase(spec: string, opts: ?SpecOptions): Spec {
+    return new Spec(spec, spec, opts);
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a dataset.
+   */
+  static forDataset(spec: string, opts: ?SpecOptions): Spec {
+    const [databaseSpec, datasetName] = splitDatabaseSpec(spec);
+    if (!datasetRe.test(datasetName)) {
+      throw new SyntaxError(`Invalid dataset ${datasetName}, must match ${datasetRe.source}`);
+    }
+
+    const specObj = new Spec(spec, databaseSpec, opts);
+    specObj._datasetName = datasetName;
+    return specObj;
+  }
+
+  /**
+   * Creates a `Spec` from a spec to a value path.
+   */
+  static forPath(spec: string, opts: ?SpecOptions): Spec {
+    const [databaseSpec, pathStr] = splitDatabaseSpec(spec);
+    const path = AbsolutePath.parse(pathStr);
+
+    const specObj = new Spec(spec, databaseSpec, opts);
+    specObj._path = path;
+    return specObj;
+  }
+
+  /**
+   * Returns the spec string.
+   */
+  spec(): string {
+    return this._spec;
+  }
+
+  /**
+   * Returns the database protocol, for example `"mem"` or `"http"`.
+   */
+  protocol(): string {
+    return this._protocol;
+  }
+
+  /**
+   * Returns the database name, e.g. `"demo.noms.io"` or `"localhost:8000"`.
+   */
+  databaseName(): string {
+    return this._databaseName;
+  }
+
+  /**
+   * Returns the `Database`.
+   */
+  database(): Database {
+    if (!this._database) {
+      this._database = this._createDatabase();
+    }
+    return this._database;
+  }
+
+  /**
+   * Returns the dataset name, for example `"sf-crime"`.
+   *
+   * Throws a `TypeError` if this spec does not have an associated value.
+   */
+  datasetName(): string {
```

Comment:
Given the convincing arguments that you make against having one parse method(i.e. ipv6 and different use cases for path vs dataset), it seems like programmers will always know what 'get' methods should be called. It's probably best to continue throwing errors so that callers don't have to constantly check for null.